### DISCUSSION TITLE
PixiJS Asset Proposal

### DISCUSSION THREAD

**Discussion by GoodBoyDigital** [pixijs/pixijs#7639] [RFC]
## PixiJS Asset Proposal 
By Mat Groves, Matt Karl

### Overview
The goal of this document is to identify the current challenges related to loading and caching assets in PixiJS and propose a solution to streamline and modernize how we resolve, load and cache resources.

### Problems
**No Single-Source of Truth**
Resource Loader maintains a per-instance cache and PixiJS maintains its own static caches ( TextureCache, BaseTextureCache). Having multiple points of caching means that assets can get loaded twice or easily blocked from garbage collection. This creates a lot of confusion for complicated applications that load and unload resources.

**Duplicate Loading Mechanism**
BaseTexture resources have implemented their own way to load resources. Whether that be loading SVGs or images as HTMLImageElements. There are redundant loading strategies between Loader and Resources that creates a maintenance headache. Generally, Resource Loader is well tested and the internal loading mechanism in BaseTexture resources is not well tested.

**Legacy Callback API Design**
Loader uses an API design that’s largely inherited or inspired from ActionScript3. Loader instances are created and callbacks are used for handing the result or errors. This OOP style has largely been replaced with Promise with APIs like fetch. It makes it easier to read code using async/await.

**Poor Separation of Concerns**
Having multiple caches fighting with each other means that there is poor separation of concerns between network management, URL resolving, asset parsing, and asset management. We need an API that more accurately manages these problems. Also, many core objects like Sprite, Texture and BaseTexture know TOO much about the cache. Convenience methods like Texture.from or storing cache keys on the objects creates complexity.

**Lack of Multithreading**
Network traffic throttles performance. Being able to use Workers and multithreading would improve performance overall for loading and parsing assets.

**Limit Resolving for Assets**
There are normal situations where you may want to request a different asset, that could be a newer less ubiquitous format like WebP over PNG, or different resolution (e.g. @2x) or if you’re using compressed-textures. Regardless, most of this responsibility is put on the developer to figure out what they need to resolve.

**Poor Caching for New Objects**
The caching system isn’t flexible enough to support caching of other objects, both high-level like Mesh or low-level like Program. These instances require their own side-car caching schemas or ref-counting. There’s not a generic way to add to the global cache.

### Introducing Assets
We’d like to introduce a new @pixi/assets package that would be a dependency of @pixi/core. This package would create three new systems to address the issues above: loader, resolver and cache.

### Asset Loader
This is a new loading API that would replace the existing Loader. Here are some of the features:

- Promise-based API
- Would maintain one internal Cache - nothing is ever loaded twice. No matter where or when an asset is requested the same promise for that asset is returned.
- One simple static loading function
- Loader plugins that test / load and transform urls into Pixi Objects (eg Textures)
- Does not touch existing caches
- Multi-threaded - will load image off thread as ImageBitmaps if possible
- Will automatically listen to texture destroy events to wipe any destroyed textures.
- Can be used independently from the resolver. Does not know WHAT to load.. Just HOW to load.

**Example**
Proposed 
```import {loadAssets} from ‘pixi.js’
const asset = await loadAssets('cool.png');
console.log(asset); // instanceof Texture```
```
Current (v6)
```
import {Loader} from ‘pixi.js’
const loader = new Loader();
loader.add('cool.png');
loader.load(() => { console.log(loader.resources[‘cool.png’]); });
```

### Asset Resolver
Resolve an asset with various inputs into a single network request. Inspired by the compressed texture manifest to resolve various texture formats. Features include:

- A separate system, knows nothing about loaders or caching
- Allows devs too tie multiple URLs per key and resolve preferred asset
- Can be used independently from the asset loader. Knows WHAT to load.. Not HOW to load.
- Will be useful for selecting various resolutions or compressed texture formats based on specified developer preferences

**Example**
```
import {Resolver} from `pixi.js`

const resolver = new Resolver();
resolver.prefer({
  resolution: [1],
  format: ['png', ‘webp’],
  priority: ['resolution', 'format']
});
resolver.add('test', [
  {
    resolution: 1,
    format: 'png',
    src: 'my-image.png'
  },
  {
    resolution: 2,
    format: 'png',
    src: 'my-image@2x.png'
  },
  {
    resolution: 1,
    format: 'webp',
    src: 'my-image.webp'
  },
  ‘my-image@2x.webp’ // sugar: infer automatically from url
]);
const url = resolver.get(‘test);
console.log(url); // it’s `'my-image.png`
```

### Asset Cache

The magic glue class that will join the above systems together. Assets will be our one stop shop for any asset type. Long-term replacing external cache based functions such as Texture.from. In the short term, Assets can run underneath the .from calls.

- Simple API for retrieving and loading assets via keys or URLs
- As assets are loaded through here all assets go through three steps:
  - URLs resolved via Resolver
  - Loaded by loadAssets
   - Added to internal Caches (e.g. Textures, Sprite Sheet textures)
- Promise based
- Suggested, paved-path API for most PixiJS users.

**Example**
```
import {Assets, Sprite} from ‘pixi.js’

Assets.texture.resolver.add('test', [
  ‘my-image.png’,
  ‘my-image@2x.png’,
  ‘my-image.webp’,
  ‘my-image@2x.webp’ 
]);

const textures = await Assets.load([‘test’, ‘other-real-image.png’]);
console.log(textures.test); // Texture `my-image.png`
console.log(textures[‘other-real-image.png’]); // Texture

// Use instead of Sprite.from
const sprite = new Sprite(textures.test);
```
#### Lazy Loading
PixiJS currently supports a style of lazy-loading Textures. You can use, implement and add to your display a Texture that is invalid and hasn’t finished loading. Should Textures continue to support lazy-loading?

The case for continuing to keep lazy-loading:
- It’s easy to get started by just doing Texture.from(‘xx.ong’) and not caring if its ready or not
- Convenient and makes easier demos
- Removing this could get messy as some textures need to ‘load’ some don’t.
- We could simply make Asset calls under the hood. But this requires the Assets functionality to reside within the core package.

The case for removing lazy-loading:
- Textures and resources can be greatly simplified, not valid / load check required in base texture classes
- Async/await creates more modern terse version of Texture.from
- Separate loading logic from rendering logic. Textures only deal with complete resources
- All textures are created via Asset

#### BaseTextureCache
Does PixiJS continue to need a BaseTextureCache? This could just be a function on the normal TextureCache - as All base textures can be retrieved by retrieving the texture with the url from the cache and accessing its base Texture.

#### New API
This proposed solution is what we have been using internally on Goodboy projects for the last year (bypassing the current Resource Loader). It is battle-tested and works really well, and I’m keen to bring these benefits to the rest of the PixiJS community.

**Comment by miltoncandelero** [comment]
I am not sure I understand how and where the `Asset Cache` lives.
In your sample implementation your store the result of the `Assets.load([‘test’, ‘other-real-image.png’]);` promise but... what if I need the loaded stuff far from the `.load` call?
Right now you can use the `loaderInstance.resources` object (like `Loader.shared.resources`) to find the loaded assets. Would that `resources` object still be there or will it be up to the end user to store their `resources`? 🤔 

---

In other notes, `Sprite.from("myTexture")` is waaaay easier than `new Sprite(Loader.shared.resources.myTexture.texture)` (or was it `.data`?)

**Reply by andrewstart** [reply]
> I am not sure I understand how and where the `Asset Cache` lives.

I believe that it would be a globally accessible, static thing, like `TextureCache` or `BaseTextureCache`, but replacing those and being more visible/accessible. Is that correct, @GoodBoyDigital?

> In other notes, `Sprite.from("myTexture")` is waaaay easier than `new Sprite(Loader.shared.resources.myTexture.texture)` (or was it `.data`?)

I tend to agree - keeping `*.from('foo.png')` as (mostly) syntactic sugar that internally uses `Assets.load('foo.png')` sounds useful to me, at least when hacking together demos or small projects.

I like where this is going, and just have a few thoughts:
1. I think you may want to elaborate more on the resolver, because when initially reading this I thought it was merely something like using a multi-extension url like `path/mySupportedAudio.{ogg,mp3}`. After rereading it, I realized that it is both that but also grants the ability to preregister all assets your app contains as early as application initialization under unique ids. I'd want to make sure that anyone considering this proposal is aware of that potentially very useful feature.
2. Does the code you've been using internally allow for plugins to easily process loaded content before accepting it/passing it off to the next plugin, or is this something we would build when adding it to PixiJS? This would be to differentiate between Spine skeletons/spritesheets/any other JSON content, for example.

**Comment by ivanpopelyshev** [comment]
I dont care about `@2x` cases much (though i have plugin that adds it in old loader) , 
but what i care about : SPINE. Spine loads several files.
1. Loading several files with specifying one or all of them
2. Using same atlas for multiple files
3. Easy way to add resource directly, its already loaded, i want loader to act like its ready!
4. integration with a hecking webpack/parcel/whatever uses "require" on resources, because people dont understand that they cant just use it, but apparantly they do it regularly.

If you have API for that, its sold to me!

**Reply by miltoncandelero** [reply]
I don't think 4 is possible without a plugin for each software.
 
(Besides, it's not a good way of doing things for pixi)

**Reply by ivanpopelyshev** [reply]
that doesnt stop people from trying it and asking me "what did i do wrong" each time :)

people also like to put tiling textures into atlases , even if its not good way of doing things in webgl
